BitList.cs

```diff
@@ -0,0 +1,1553 @@
+// By Lorenzo Lotti
+
+using System;
+using System.Linq
+using System.IO;
+using System.Collections;
+using System.Collections.Generic;
+
+/// <summary>
+/// Exception generated calling a method when the BitList instance is empty.
+/// </summary>
+public class EmptyInstanceException : Exception { }
+
+/// <summary>
+/// Exception generated when a boolean operation fails.
+/// </summary>
+public class BooleanAlgebraException : Exception { }
+
+/// <summary>
+/// Exception generated when a BitList instance fails to write or read a file.
+/// </summary>
+public class FileException : Exception { }
+
+/// <summary>
+/// Exception generated when a BitList instance fails to read a file.
+/// </summary>
+public class FileReadingException : FileException { }
+
+/// <summary>
+/// Exception generated when a BitList instance fails to write a file.
+/// </summary>
+public class FileWritingException : FileException { }
+
+/// <summary>
+/// Exception generated when a BitList instance fails the conversion to another type.
+/// </summary>
+public class ConversionErrorException : Exception { }
+
+/// <summary>
+/// A list of bits.
+/// </summary>
+/// <exception cref="EmptyInstanceException" />
+/// <exception cref="BooleanAlgebraException" />
+/// <exception cref="FileReadingException" />
+/// <exception cref="FileWritingException" />
+/// <exception cref="ConversionErrorException" />
+public sealed class BitList :
+    IEnumerable<bool>,
+    ICollection<bool>,
+    IList<bool>,
+    ICloneable,
+    IEquatable<BitList>
+{
+    private BitArray bits;
+    bool ICollection<bool>.IsReadOnly { get => false; }
+
+    /// <summary>
+    /// This property return true if the instance has no values.
+    /// </summary>
+    public bool IsEmpty { get => bits.Equals(null) || bits.Equals(new BitArray(0)); }
+
+    /// <summary>
+    /// The amount of bits of the instance.
+    /// </summary>
+    public int Count { get => IsEmpty ? 0 : bits.Length; }
+
+    /// <summary>
+    /// Indexer.
+    /// </summary>
+    public bool this[int index]
+    {
+        get => Get(index);
+        set => Set(index, value);
+    }
+
+    /// <summary>
+    /// Indexer.
+    /// </summary>
+    /// <param name="index">Start index.</param>
+    /// <param name="count">Count of bits.</param>
+    /// <exception cref="ArgumentOutOfRangeException" />
+    public BitList this[int index, int count]
+    {
+        get => Get(index, count);
+        set
+        {
+            if (count == value.Count)
+                Set(index, value);
+            else
+                throw new ArgumentOutOfRangeException();
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use a constructor overloads that allows a more meaningful exception message to be provided.](https://app.codacy.com/manual/siriak/C-Sharp/pullRequest?prid=5545753)